### ISSUE TITLE
line continuation in preprocessor in documented text (Origin: bugzilla #684050)

### ISSUE THREAD

**Issue by doxygen** [doxygen/doxygen#4887]
status NEW severity *normal* in component *general* for *---*
Reported in version *1.8.2* on platform *Other*
Assigned to: Dimitri van Heesch

On 2012-09-14 21:46:52 +0000, Matthew Woehlke wrote:

    I am trying to define some standard methods with common documentation, like
    so:
    
    #define MY_STANDARD_METHODS(name) \
      /** Brief description of name. \
          Detailed description of name. \
       */ \
      void my_standard_method()
    
    This *almost* works, except that doxygen emits the line continuation (\)
    characters verbatim in the documentation, which does not seem correct.
    
    Is this a bug (seems like it...)? Is there a way around this that does not
    require using ALIASES? (I have a single doxyfile template I use across
    multiple libraries, and would prefer to keep this sort of thing per-
    library.)

On 2012-09-14 21:52:33 +0000, Matthew Woehlke wrote:

    So, to answer part of my own question...
    
    #define MY_STANDARD_METHODS(name) \
      /** Brief description of name.  Detailed description of name.  */ \
      void my_standard_method()
    
    ...which is a fascinating hack, but still wondering if it should be
    necessary.

On 2017-10-14 16:46:36 +0000, albert wrote:

    In https://stackoverflow.com/questions/46744573/does-doxygen-have-a-line-
    continuation-feature of user AcId on stackoverflow it was mentioned that in
    the following example:
    /**
     * \file mwe.cpp
     * 
     * \details
     * MWE::MWE(int a, int b, int c)
     * MWE::MWE(int a,
     *           int b, int c)
     */
    
    
    class MWE
    {
        public:
            /** 
             * \brief constructor with one parameter
             */
            MWE(int a);
    
            /** 
             * \brief constructor with three parameters
             **/
            MWE(int a, int b, int c);
    }
    
    When generating the documentation, the first link (MWE::func(int a, int b,
    int c)) correctly points to the constructor taking three integer parameters.
    However, the second link, where there is a line-break, points to the
    constructor taking only a single integer parameter (also the parameter list
    does not become part of the link, only the function name).
    
    A "solution" for this was, similar to the solution of Matthew Woehlke:
    
    /**
     * \file mwe.cpp
     * 
     * \details
     * MWE::MWE(int a, int b, int c)
     * MWE::MWE(int a,          int b, int c)
     */
    
    
    class MWE
    {
        public:
            /** 
             * \brief constructor with one parameter
             */
            MWE(int a);
    
            /** 
             * \brief constructor with three parameters
             **/
            MWE(int a, int b, int c);
    }

**Comment by craffael** [comment]
Would be nice if this bug could be resolved some time...?

**Comment by karthikeyann** [comment]
`@copydoc` does not work if function signature spans to multiple lines.
this trick does not work for copydoc

**Comment by albert-github** [comment]
@karthikeyann 
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.
- Please also specify the **full** doxygen version used (`doxygen -v`).

**Comment by karthikeyann** [comment]
Generated doxygen configuration using `doxygen -g` and modified `EXTRACT_ALL = YES` and `INPUT = .`
- [test_doxy.zip](https://github.com/doxygen/doxygen/files/7276596/test_doxy.zip)
- Generated using doxygen 1.8.18 (13b3d330f24ff38b30d8b58ebd2ded6ec1ab7a85*)

<details><summary>test.hpp</summary>
<p>

```

#include <algorithm>
#include <utility>
#include <vector>

/**
 * @brief test function 1 brief
 *
 * test function 1 detailed description
 *
 * @param arg1 argument 1
 * @param arg2 argument 2
 * @param arg3 argument 3
 * @param arg4 argument 4
 * @param arg5 argument 5
 * @return vector of bool uninitialized values.
 */
std::vector<bool> test_function1(std::vector<int> const& arg1,
                                 std::vector<double> const& arg2,
                                 std::pair<unsigned, unsigned> arg3,
                                 std::vector<std::pair<unsigned long, unsigned short>> const& arg4,
                                 std::vector<bool> const& arg5);

/**
 * @copydoc test_function1(std::vector<int> const& arg1,
 *                         std::vector<double> const& arg2,
 *                         std::pair<unsigned, unsigned> arg3,
 *                         std::vector<std::pair<unsigned long, unsigned short>> const& arg4,
 *                         std::vector<bool> const& arg5)
 *
 * @param arg6 argument 6 int
 */
std::vector<bool> test_function2(std::vector<int> const& arg1,
                                 std::vector<unsigned> const& arg2,
                                 std::pair<float, double> arg3,
                                 std::vector<std::pair<unsigned long, unsigned short>> const& arg4,
                                 std::vector<bool> const& arg5,
                                 int arg6);

/**
 * @brief  @copybrief test_function1(std::vector<int> const& arg1,
 *                                   std::vector<double> const& arg2,
 *                                   std::pair<unsigned, unsigned> arg3,
 *                                   std::vector<std::pair<unsigned long, unsigned short>> const&
 * arg4, std::vector<bool> const& arg5)
 * @details  @copydetails test_function1(std::vector<int> const& arg1,
 *                                       std::vector<double> const& arg2,
 *                                       std::pair<unsigned, unsigned> arg3,
 *                                       std::vector<std::pair<unsigned long, unsigned short>>
 * const& arg4, std::vector<bool> const& arg5)
 * @param arg6 argument 6 double
 */
std::vector<bool> test_function3(std::vector<int> const& arg1,
                                 std::vector<unsigned> const& arg2,
                                 std::pair<float, double> arg3,
                                 std::vector<std::pair<unsigned long, unsigned short>> const& arg4,
                                 std::vector<bool> const& arg5,
                                 double arg6);

```
</p>
</details>

```
Generating code for file test.hpp...
/home/test_doxy/test.hpp:25: warning: @copybrief or @copydoc target 'test_function1(std::vector<int> const& arg1,' not found
Generating file documentation...
Generating docs for file test.hpp...
/home/test_doxy/test.hpp:25: warning: @copybrief or @copydoc target 'test_function1(std::vector<int> const& arg1,' not found
/home/test_doxy/test.hpp:25: warning: @copybrief or @copydoc target 'test_function1(std::vector<int> const& arg1,' not found
```
But it generated documentation for `test_function2`, and `test_function3`.
`test_function2` documentation is missing brief.

**Comment by albert-github** [comment]
@karthikeyann 
Indeed a bit of a strange situation, the `\copybrief` and `\copydetails` do handle a split line properly but `\copydoc` doesn't.
At the moment `\copydoc` terminates at the end of a line, as a **workaround** you can place the entire `\copydoc` on one line.
So 
```
 * @copydoc test_function1(std::vector<int> const& arg1,std::vector<double> const& arg2,std::pair<unsigned, unsigned> arg3,std::vector<std::pair<unsigned long, unsigned short>> const& arg4,std::vector<bool> const& arg5)
```
instead of:
```
 * @copydoc test_function1(std::vector<int> const& arg1,
 *                         std::vector<double> const& arg2,
 *                         std::pair<unsigned, unsigned> arg3,
 *                         std::vector<std::pair<unsigned long, unsigned short>> const& arg4,
 *                         std::vector<bool> const& arg5)
```

**Comment by albert-github** [comment]
This issue now contains 3 problems:
1. [x] the line continuation issue in the `#define`. In the doxygen version 1.8.2 (till 1.8.10) I can reproduce this problem, in the current version this is not present anymore, so I think this problem has been solved in the mean time
2. [ ] The problem of split line when using autolinking, this problem is still present
3. [x] the `\copydoc` problem as signaled by @karthikeyann in https://github.com/doxygen/doxygen/issues/4887#issuecomment-933258137, for this I've just pushed a proposed patch, pull request #8812

**Comment by albert-github** [comment]
The problem of split line when using autolinking, I've just pushed a proposed patch, pull request #8814

**Comment by karthikeyann** [comment]
@albert-github The workaround is good. Longer single line could not be used when `clang-format` is used to format the code which folds the longer lines.

**Comment by albert-github** [comment]
The functionality of the  proposed pull request #8812 has been made, in a slightly different way, available through #8977

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by karthikeyann** [comment]
Thank you @albert-github 👍

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.3.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
